standalone-metastore/metastore-server/src/main/sql/mssql/hive-schema-4.0.0-beta-2.mssql.sql

```diff
@@ -661,6 +661,8 @@ ALTER TABLE NOTIFICATION_LOG ADD CONSTRAINT NOTIFICATION_LOG_PK PRIMARY KEY (NL_
 
 CREATE UNIQUE INDEX NOTIFICATION_LOG_EVENT_ID ON NOTIFICATION_LOG (EVENT_ID);
 
+CREATE UNIQUE INDEX NOTIFICATION_LOG_UNIQUE_DB ON NOTIFICATION_LOG (DB_NAME, EVENT_ID);
```

Comment:
As we have already indexed on `EVENT_ID` for this table, do we need to create a composite index on (DB_NAME, EVENT_ID)?

Reply:
Both indices are required. There can be two different types of frequent queries in this table.

1. provided eventId, fetch all the events greater than the eventId. -- Will use index with event_id 
2. provided eventId, dbName, table, fetch all the events of this table greater than the eventId. -- Will use index with db_name and event_id

Reply:
I'm wondering for some back db, for the second case: event_id > 10 and db_name = 'default'
 1. use index with event_id to filter the matched rows (event_id > 10)
 2. on the result of the first step, apply the index with db_name (db_name = 'default')
if the event_id > 10 is only return one row from 1000 rows, then this is better than using the index with db_name and event_id.
As the API is aim to fetch the latest event in most cases, so I think this is common in real production.

Reply:
Any DBMS would use one and only one index to query the result. So depending on the query on the notification log table, it has to use index 1 or index 2.

Reply:
I agree with @dengzhhu653 
Optimizer would check the potential indexes available and decides to choose one(could be single index or composite index) among them or not to choose any index based on various factors. Since we have index with single column(`EVENT_ID`), in query having where clause with eventId + may be dbName + may be table name etc, optimizer detects the eventId index as potential candidate. Have captured explain queries on mysql for index used and not used cases.
Index used:
```
mysql> explain select * from NOTIFICATION_LOG where db_name='mydb' and event_id>8;
+----+-------------+------------------+------------+-------+---------------------------+---------------------------+---------+------+------+----------+------------------------------------+
| id | select_type | table            | partitions | type  | possible_keys             | key                       | key_len | ref  | rows | filtered | Extra                              |
+----+-------------+------------------+------------+-------+---------------------------+---------------------------+---------+------+------+----------+------------------------------------+
|  1 | SIMPLE      | NOTIFICATION_LOG | NULL       | range | NOTIFICATION_LOG_EVENT_ID | NOTIFICATION_LOG_EVENT_ID | 8       | NULL |    3 |    10.00 | Using index condition; Using where |
+----+-------------+------------------+------------+-------+---------------------------+---------------------------+---------+------+------+----------+------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> explain select * from NOTIFICATION_LOG where db_name='mydb' and tbl_name='stage' and event_id>8;
+----+-------------+------------------+------------+-------+---------------------------+---------------------------+---------+------+------+----------+------------------------------------+
| id | select_type | table            | partitions | type  | possible_keys             | key                       | key_len | ref  | rows | filtered | Extra                              |
+----+-------------+------------------+------------+-------+---------------------------+---------------------------+---------+------+------+----------+------------------------------------+
|  1 | SIMPLE      | NOTIFICATION_LOG | NULL       | range | NOTIFICATION_LOG_EVENT_ID | NOTIFICATION_LOG_EVENT_ID | 8       | NULL |    3 |     9.09 | Using index condition; Using where |
+----+-------------+------------------+------------+-------+---------------------------+---------------------------+---------+------+------+----------+------------------------------------+
1 row in set, 1 warning (0.00 sec)

```
Index doesn't seem to be used here:
```
mysql> explain select * from NOTIFICATION_LOG where event_id>12 and db_name='mydb';
+----+-------------+------------------+------------+------+---------------------------+------+---------+------+------+----------+-------------+
| id | select_type | table            | partitions | type | possible_keys             | key  | key_len | ref  | rows | filtered | Extra       |
+----+-------------+------------------+------------+------+---------------------------+------+---------+------+------+----------+-------------+
|  1 | SIMPLE      | NOTIFICATION_LOG | NULL       | ALL  | NOTIFICATION_LOG_EVENT_ID | NULL | NULL    | NULL |   29 |     5.86 | Using where |
+----+-------------+------------------+------------+------+---------------------------+------+---------+------+------+----------+-------------+
1 row in set, 1 warning (0.00 sec)
```
Have found some reference from mysql but this seem to be not very specific to mysql.
https://dev.mysql.com/doc/refman/8.0/en/multiple-column-indexes.html#:~:text=If%20the%20table%20has%20a%20multiple,of%20(col1%2C%20col2%2C%20col3).
https://dev.mysql.com/doc/refman/8.0/en/index-condition-pushdown-optimization.html#:~:text=Get%20the%20next%20row%27s,on%20the%20test%20result.

Some dbs seem to provide a way to specify index hints in query explicilty to always use an index. 

If we add the dbname + eventId composite index, index updation/maintence overhead may be high because index data need to be sorted by dbname first and then by eventIds within the db on the huge notification log? It may improve the query performance though.

May be good to compare the tradeoffs between using the eventId based index vs dbname + eventId composite index ?

Reply:
@VenuReddy2103 - Good finding. The index I introduced is only used if there is event_id value is absolute
`mysql> explain analyze select * from NOTIFICATION_LOG where db_name='mydb';
+-----------------------------------------------------------------------------------------------------------------------------------------------------+
| EXPLAIN                                                                                                                                             |
+-----------------------------------------------------------------------------------------------------------------------------------------------------+
| -> Index lookup on NOTIFICATION_LOG using NOTIFICATION_LOG_UNIQUE_DB (DB_NAME='mydb')  (cost=0.9 rows=4) (actual time=0.276..0.738 rows=4 loops=1)
 |
+-----------------------------------------------------------------------------------------------------------------------------------------------------+`
But the real use case query would definitely have a range of conditions on the event_id (eg: event_id>100), but apparently the first index is being used 'NOTIFICATION_LOG_EVENT_ID' instead of 'NOTIFICATION_LOG_UNIQUE_DB'. This is something not desired.
`mysql> explain analyze select * from NOTIFICATION_LOG where db_name='mydb' and (tbl_name='tb1' or tbl_name='tb2') and event_id>1;
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| EXPLAIN                                                                                                                                                                                                                                                                                                                                                                                                                      |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| -> Filter: ((notification_log.DB_NAME = 'mydb') and ((notification_log.TBL_NAME = 'tb1') or (notification_log.TBL_NAME = 'tb2')))  (cost=1.61 rows=1.31) (actual time=0.0784..0.0969 rows=2 loops=1)
    -> Index range scan on NOTIFICATION_LOG using NOTIFICATION_LOG_EVENT_ID over (1 < EVENT_ID), with index condition: (notification_log.EVENT_ID > 1)  (cost=1.61 rows=3) (actual time=0.0731..0.0902 rows=3 loops=1)
 |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+`
So in summary the second index I intend to introduce has no effect. There is no point in forcing the DBMS to use the index because it is inefficient in reading the results and has an overhead to maintain the second index. I would rather remove the newly introduced index. Thanks